rust_src/src/strings.rs

```diff
@@ -33,24 +33,13 @@ fn string_bytes(string: LispObject) -> LispObject {
 /// Case is significant, but text properties are ignored.
 /// Symbols are also allowed; their print names are used instead.
 #[lisp_fn]
-pub fn string_equal(mut s1: LispObject, mut s2: LispObject) -> LispObject {
-    if s1.is_symbol() {
-        s1 = LispObject::from(unsafe { SYMBOL_NAME(s1.to_raw()) });
-    }
-    if s2.is_symbol() {
-        s2 = LispObject::from(unsafe { SYMBOL_NAME(s2.to_raw()) });
-    }
-    let mut s1 = s1.as_string_or_error();
-    let mut s2 = s2.as_string_or_error();
+pub fn string_equal(s1: LispObject, s2: LispObject) -> LispObject {
+    let s1 = LispObject::symbol_or_string_as_string(s1);
+    let s2 = LispObject::symbol_or_string_as_string(s2);
 
     LispObject::from_bool(
-        s1.len_chars() == s2.len_chars() && s1.len_bytes() == s2.len_bytes() && unsafe {
-            libc::memcmp(
-                s1.data_ptr() as *mut c_void,
-                s2.data_ptr() as *mut c_void,
-                s1.len_bytes() as usize,
-            ) == 0
-        },
+        s1.len_chars() == s2.len_chars() && s1.len_bytes() == s2.len_bytes()
+            && s1.chars().zip(s2.chars()).all(|(c1, c2)| c1 == c2),
```

Comment:
That translation decodes the individual characters, which is much slower than the `memcmp`, and unnecessary here. But we have `as_slice`, so you should be able to compare the two slices without requiring `unsafe`.

BTW, this would otherwise be written with `Iterator::eq`.

Reply:
Also, comparing slices will make the `len_bytes()` check unnecessary.

Reply:
I was not sure I could trust Iterator::eq, I intended to test it.

Slice comparison does make more sense. I started off mimicking `string-lessp` and when I realized I did not need to I did not come back to this. Thanks for the catch.